Fix some issues withe the calculation of `__fp_one`

miscco opened the pull request:
The problem occured with floating point types without an implicit bit like `long double` which however takes a different code path.

The ther issue was that we did not properly cast the integer of the bias to the storage type, which could again could lead to issues with types that have more mantissa bits than `int`.

However, currently the only supported types that satisfy this are builtin types, which again take a different code path.

Still worth fixing though

miscco committed:
Fix some issues withe the calculation of `__fp_one`

The problem occured with floating point types without an implicit bit like `long double` which however takes a different code path.

The ther issue was that we did not properly cast the integer of the bias to the storage type, which could again could lead to issues with types that have more mantissa bits than `int`.

However, currently the only supported types that satisfy this are builtin types, which again take a different code path.

Still worth fixing though

davebayer commented on libcudacxx/include/cuda/std/__floating_point/constants.h:
```diff
@@ -193,7 +193,17 @@ template <class _Tp>
+  }
+  else
+  {
+    constexpr __fp_storage_t<_Fmt> implicit_bit = __fp_storage_t<_Fmt>{1} << (__fp_mant_nbits_v<fmt> - 1);
+    return static_cast<__fp_storage_t<_Fmt>>(
+      (static_cast<__fp_storage_t<_Fmt>>(__fp_exp_bias_v<_Fmt>) << __fp_mant_nbits_v<_Fmt>) &implicit_bit);
```
```suggestion
    constexpr __fp_storage_t<_Fmt> __implicit_bit = __fp_storage_t<_Fmt>{1} << (__fp_mant_nbits_v<fmt> - 1);
    return static_cast<__fp_storage_t<_Fmt>>(
      (static_cast<__fp_storage_t<_Fmt>>(__fp_exp_bias_v<_Fmt>) << __fp_mant_nbits_v<_Fmt>) | __implicit_bit);
```

miscco replied:
Someone over here needs a coffee

davebayer reviewed (changes requested):
Can we add tests for both `__fp_one` and `__fp_zero`:
1. test the return value of `__fp_constant</*format*/>()` against the reference storage representation
2. test the return value of `__fp_constant</*type*/>()` against the `__fp_constant<__fp_format_of_v</*type*/>>()`

miscco committed:
Thats ugly

miscco committed:
Add tests

miscco committed:
Avoid gcc and clang issue

Booth compilers have issues with a constexpr variable in a discarded statement that would be invalid. :shrug:
